### ISSUE TITLE
Support for additional distributions in the Standard API

### ISSUE THREAD

**Issue by vinopm** [ScottishCovidResponse/SCRCIssueTracking#671]
Discussed with @mrow84 that we would need to add support for 'exponential' and 'uniform' distributions in all Standard API implementations. 

We also need to update the Standard API spec with more information about which distributions are supported.

I believe the complete list of distributions we would support are:

- Gamma
- Exponential
- Linear

- [x] Update API specification document

- [x] Python implementation

- [ ] Java implementation

- [ ] C++ implementation

- [ ] Julia implementation

**Comment by vinopm** [comment]
@mrow84 In the Contact Tracing Model, there is an input distirbution where we pass in a list of bins and a list of their respective weights. The model creates an enumerated integer distribution out of this. Can we add in support for this type of distribution as well?

Something like this?

```
[population-ages]
type = "distribution"
distribution = "enumerated"
bins = [0, 15, 25, 55, 65, 90]
weights = [0.1759, 0.1171, 0.4029, 0.1222, 0.1819]
```

**Comment by mrow84** [comment]
> @mrow84 In the Contact Tracing Model, there is an input distirbution where we pass in a list of bins and a list of their respective weights. The model creates an enumerated integer distribution out of this. Can we add in support for this type of distribution as well?
> 
> Something like this?
> 
> ```
> [population-ages]
> type = "distribution"
> distribution = "enumerated"
> bins = [0, 15, 25, 55, 65, 90]
> weights = [0.1759, 0.1171, 0.4029, 0.1222, 0.1819]
> ```

Yep, I think I would call this a [categorical distribution](https://en.wikipedia.org/wiki/Categorical_distribution).

**Comment by vinopm** [comment]
> > @mrow84 In the Contact Tracing Model, there is an input distirbution where we pass in a list of bins and a list of their respective weights. The model creates an enumerated integer distribution out of this. Can we add in support for this type of distribution as well?
> > Something like this?
> > ```
> > [population-ages]
> > type = "distribution"
> > distribution = "enumerated"
> > bins = [0, 15, 25, 55, 65, 90]
> > weights = [0.1759, 0.1171, 0.4029, 0.1222, 0.1819]
> > ```
> 
> Yep, I think I would call this a [categorical distribution](https://en.wikipedia.org/wiki/Categorical_distribution).

Ok, so we will have support for this format:

```
[population-ages]
type = "distribution"
distribution = "categorical"
bins = [0, 15, 25, 55, 65, 90]
weights = [0.1759, 0.1171, 0.4029, 0.1222, 0.1819]
```

**Comment by mrow84** [comment]
The only thing I wonder is if we might want the categories to be strings rather than numbers, and then parse them when it is deemed appropriate.

**Comment by vinopm** [comment]
@mrow84 the categories refer to ranges:

i.e. Age 0-15 -> 0.1759 probability

**Comment by mrow84** [comment]
In some sense that makes me feel like a string may be even more appropriate, in that you could encode the range more explicitly - they do something like that [in simple network sim](https://github.com/ScottishCovidResponse/simple_network_sim/blob/master/simple_network_sim/loaders.py#L303). I am happy for you to leave the range stuff, but I do think that if the string conversion isn't too difficult that it would be a positive, because it is useful to be able to form discrete distributions over more arbitrary categories.

**Comment by vinopm** [comment]
@mrow84 I agree with you. Will require quite a bit of work to support the string range format, i'll add that as a TODO, but let's keep this format for now as a first step. Would that be ok?

**Comment by mrow84** [comment]
I have been going through the distributions trying to come up with standardised parameterisations. This is what I have now, with links to wikipedia for parameterisation references, and includes both the distributions required by the java models and the EERA model (@kzscisoft / @peter-t-fox). I realise that it may in some circumstances require change file contents, so please let me know if this is too much of a drag, but I think we may already have some differences anyway, so someone is going to have to change something!

### [gamma](https://en.wikipedia.org/wiki/Gamma_distribution#Characterization_using_shape_k_and_scale_%CE%B8)
* `k`: shape
* `theta`: scale
### [normal](https://en.wikipedia.org/wiki/Normal_distribution#General_normal_distribution)
* `mu`: mean
* `sigma`: standard deviation
### [uniform](https://en.wikipedia.org/wiki/Uniform_distribution_(continuous)#Probability_density_function)
* `a`: lower bound
* `b`: upper bound
### [poisson](https://en.wikipedia.org/wiki/Poisson_distribution#Probability_mass_function)
* `lambda`: rate
### [exponential](https://en.wikipedia.org/wiki/Exponential_distribution#Probability_density_function)
* `lambda`: rate
### [categorical (non-standard)](https://en.wikipedia.org/wiki/Categorical_distribution#Formulating_distributions)
* `bins`: category labels
* `weights`: category probabilities
### [beta](https://en.wikipedia.org/wiki/Beta_distribution#Probability_density_function)
* `alpha`: shape
* `beta`: shape

**Comment by mrow84** [comment]
Also adding

### [binomial](https://en.wikipedia.org/wiki/Binomial_distribution#Probability_mass_function)
* `n`: trials
* `p`: probability of success
### [multinomial](https://en.wikipedia.org/wiki/Multinomial_distribution#Probability_mass_function)
* `n`: trials
* `p`: probabilities of success
